src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,674 @@
+  else do
+    nop
+
+submitVote
+  :: forall store pm. (NiceParameter pm, StorageC store pm)
+  => '[VoteParam, store] :-> '[store]
+submitVote = do
+  duupX @2; duupX @2
+  toField #vProposalKey
+
+  stGet #sProposals
+  ifSome nop (failCustom_ #pROPOSAL_NOT_EXIST)
+  stackType @[Proposal pm, VoteParam, store]
+
+  dig @2
+  duupX @3; toField #vVoteAmount; sender; swap
+  freeze; dig @2; dig @2
+  stackType @[Proposal pm, VoteParam, store]
+
+  constructT @(Proposal pm)
+    ( fieldCtor $ do
+        -- | Update upvote
+        duupX @2;
+        stackType @[VoteParam, Proposal pm, VoteParam, store]
+        toField #vVoteType;
+        stackType @[Bool, Proposal pm, VoteParam, store]
+        if Holds then do
+          duupX @2
+          stackType @[VoteParam, Proposal pm, VoteParam, store]
+          toField #vVoteAmount
+          duupX @2
+          stackType @[Proposal pm, Natural, Proposal pm, VoteParam, store]
+          toField #pUpvotes
+          add
+        else do
+          stackType @[Proposal pm, VoteParam, store]
+          getField #pUpvotes
+    , fieldCtor $ do
+        -- | Update downvote
+        duupX @2
+        stackType @[VoteParam, Proposal pm, VoteParam, store]
+        toField #vVoteType;
+        stackType @[Bool, Proposal pm, VoteParam, store]
+        if Holds then do
+          stackType @[Proposal pm, VoteParam, store]
+          getField #pDownvotes
+        else do
+          duupX @2
+          stackType @[VoteParam, Proposal pm, VoteParam, store]
+          toField #vVoteAmount
+          duupX @2
+          stackType @[Proposal pm, Natural, Proposal pm, VoteParam, store]
+          toField #pDownvotes
+          add
+
+    , fieldCtor $ getField #pStartDate
+    , fieldCtor $ getField #pMetadata
+    , fieldCtor $ getField #pProposer
+    , fieldCtor $ getField #pProposerFrozenToken
+    , fieldCtor $ do
+        duupX @2; toField #vVoteAmount
+        sender; pair
+        dip $ getField #pVoters
+        stackType @[(Address, Natural), [(Address, Natural)], Proposal pm, VoteParam, store]
+        cons
+    )
+
+  stackType @[Proposal pm, Proposal pm, VoteParam, store]
+  dip drop
+  stackType @[Proposal pm, VoteParam, store]
+
+  some
+  swap; toField #vProposalKey
+  stackType @[ProposalKey, Maybe (Proposal pm), store]
+  stUpdate #sProposals
+
+checkVoteLimitReached
+  :: forall store pm. (IsoValue pm, StorageC store pm)
+  => Config pm -> '[Proposal pm, VoteParam, store] :-> '[Proposal pm, VoteParam, store]
+checkVoteLimitReached Config{..} = do
+  duupX @2; duupX @2
+
+  getField #pUpvotes; dip (toField #pDownvotes); add
+  dip $ toField #vVoteAmount
+  add; toNamed #newAmount
+
+  push cMaxVotes; toNamed #maxAmount
+  if #maxAmount <. #newAmount then
+    failCustom_ #mAX_VOTES_REACHED
+  else nop
+
+-- | Vote
+vote
+  :: forall store pm. (StorageC store pm, NiceParameter pm)
+  => Config pm -> Entrypoint [VoteParam] store
+vote config = do
+  iter $ do
+    duupX @2
+    duupX @2
```

Comment:
Minor: you could use `dupTop2` here, it should be available in the `lorentz` version we use here (if it's not, ignore my comment).

Reply:
`dupTop2` is indeed available. I've replaced all `duupX @2; duupX @2` with `dupTop2` instead.